### ISSUE TITLE
Doubts about the loss_cls and accuracy_cls calculation

### ISSUE THREAD

**Issue by QiaoranC** [roytseng-tw/Detectron.pytorch#36]
Hi:
I have some doubts in the evaluation of loss_cls and accuracy_cls in  function of **fast_rcnn_losses in lib/modling/fast_rcnn_heads.py**.
Based on my understanding,  the following calculation seems assume cls_score and rois_label have the same length and matching order. Like pred [0,1,1], lable [0,1,2] (just the idea). 
But the real is more like pred [0,1,2,3], label [0,1,2] (pred length may more or less And the order may not match).
Based on my experience in matterport's mask rcnn. Before calculate the class loss and accuracy, there is operation will matching the pred and label in order based on the nearest box. Basically, it found the nearest pred bbox as the 'right' pred for one label box (make sense). 

I didnot found some operation in the code yet, i guess i ignore or misunderstand something (new to the mask rcnn/faster rcnn). 
So my real question is **how do you make sure the pred class and label class matching before calculate the loss/accuracy?** 
thanks. 

```
def fast_rcnn_losses(cls_score, bbox_pred, label_int32, bbox_targets,
                     bbox_inside_weights, bbox_outside_weights):
    device_id = cls_score.get_device()
    rois_label = Variable(torch.from_numpy(label_int32.astype('int64'))).cuda(device_id)
    loss_cls = F.cross_entropy(cls_score, rois_label)
    ........
    # class accuracy
    cls_preds = cls_score.max(dim=1)[1].type_as(rois_label)
    accuracy_cls = cls_preds.eq(rois_label).float().mean(dim=0)

    return loss_cls, loss_bbox, accuracy_cls
```

**Comment by roytseng-tw** [comment]
Check out the function `_sample_rois` [here](https://github.com/roytseng-tw/Detectron.pytorch/blob/master/lib/roi_data/fast_rcnn.py#L1291)

**Comment by QiaoranC** [comment]
Thanks for the pointing. 
One more silly/(time saver for me) question:
`_sample_rois`  in `roi_data.fast_rcnn` seems only imported by collect_and_distribute_fpn_rpn_proposals.py
But then `collect_and_distribute_fpn_rpn_proposals.py` or  `fast_rcnn.py` not imported by  `fast_rcnn_heads.py` or `model_builder.py`  or other palce(not found). 

So How the `_sample_rois` been really connected in `fast_rcnn_losses`?

I am asking this in details because i finally want calculate the precision,recall,f1 for the each class on my dataset during training.

**Comment by QiaoranC** [comment]
found it
